src/solvers/refinement/string_refinement.cpp

```diff
@@ -1085,21 +1096,49 @@ static exprt negation_of_constraint(const string_constraintt &axiom)
   return negaxiom;
 }
 
+/// Result of the solver `supert` should not be interpreted literally for char
+/// arrays as not all indices are present in the index set.
+/// In the given expression, we populate arrays at the indices for which the
+/// solver has no constraint by copying values to the left.
+/// For example an expression `ARRAY_OF(0) WITH [1:=2] WITH [4:=3]` would
+/// be interpreted as `{ 2, 2, 3, 3, 3}`.
+/// \param expr: expression to interpret
+/// \param string_max_length: maximum size of arrays to consider
+/// \return the interpreted expression
+exprt concretize_arrays_in_expression(
```

Comment:
This function is probably a good place to use @LAJW's new tree-walking copy-minimising expression iterator: https://github.com/diffblue/cbmc/blob/test-gen-support/src/util/expr_iterator.h

Basically it walks over the tree using const iterators, then when you find an ID_index node with an ID_with child you call .mutate() to get a writable iterator for that particular subexpression, cf. using non-const `operands()` below which will break all sharing even when there are no changes necessary.